### DISCUSSION TITLE
Python Wrangling

### DISCUSSION THREAD

**Discussion by ELITSZORD** [teamalgoritma/community#812] [Q&A] [answered]
Dear Algoritma Team,

![case1 1](https://github.com/teamalgoritma/community/assets/114390619/11a54850-c7dd-4e51-a181-1b5c13ff1362)

Image 1 di atas.

![case1 1 2](https://github.com/teamalgoritma/community/assets/114390619/7d36d3e1-e2b5-4086-b3ed-d0c575e908fc)

Image 2 di atas

Q1 : Bagaimana image 2 bisa dimasukkan ke dalam image 1 dimana menggantikan nilai NaN yang menyesuaikan unique identifier(Id) pada python?

![Case 2 2](https://github.com/teamalgoritma/community/assets/114390619/099e72d3-0e6c-41f7-a45b-65f8acefb72c)
 Q2 : Bagaimana table tersebut bisa dipindah ke bawah table sebelumnya?

Kondisi dimana data tersebut terkonek ke API dan update setiap harinya.

Mohon bantuannya temen2, terimakasih atas perhatiannya dan kerjasamanya.
Terimakasih

Best Regards,

Faisal

**Comment by kinantytasyao** [comment]
Dear @ELITSZORD 

Pertanyaan Anda telah kami terima dan sedang kami usahakan terkait solusinya. Mohon kesediaannya untuk menunggu, terima kasih.

Best regards,
Team Algoritma

**Comment by arkanayudhistira** [comment]
Dear @ELITSZORD,

Berikut adalah solusi untuk menjawab kedua problem tersebut:

# Q1

Kami memisalkan variabel yang kita gunakan adalah

- `df_missing` = dataframe dengan nilai NaN pada kolom `"Actual"`
- `df_new` = dataframe baru yang memiliki nilai `"Actual"`

**for loop** dilakukan untuk mengiterasi seluruh baris pada `df_new`, sehingga `ID` nya dapat disimpan kemudian dicocokkan dengan `ID` pada `df_missing` menggunakan `loc` dan **conditional subsetting**. Nilai `Actual` pada baris yang `ID` nya cocok akan direplace. Code nya akan menjadi seperti berikut:

1. Jika kolom `"ID"` merupakan index dari dataframe :
  ```
  for id in df_new.index:
      df_missing.loc[id, 'Actual'] = df_new.loc[id, 'Actual']
  ```
2. Jika kolom `"ID"` bukan merupakan index dari dataframe (hanya kolom biasa):
```
 for id in df_new['ID']:
    df_missing.loc[df_missing['ID'] == id, 'Actual'] = df_new.loc[df_new['ID'] == id, 'Actual']
```

# Q2

Untuk menggabung baris-baris dari 2 atau lebih dataframe dengan kolom-kolom yang sama, anda dapat menggunakan `pd.concat()` dengan contoh seperti berikut:

```
df_concat = pd.concat([df1, df2])
```

Semoga membantu!

**Reply by ELITSZORD** [reply]
Misal sudah ada output df baru kemudian mau di save ke specific nama file excel & specific nama worksheet itu gimana ya mas @arkanayudhistira ? seperti overwrite, tapi tanpa merubah format tipe data df baru.

Kemudian ketika saya read file excel menggunakan `pd.read_excel(path,sheet_name='kota')` kenapa format datanya berubah khususnya date dimana di excel saya set format datanya dd/mm/yy tp pas di read pake pandas format kolom date berubah menjadi yy-mm-dd dan terkadang ada tambahan hh:mm:ss? pengennya ketika kita preview format data pada excel sama dengan ketikan dibuka pake pandas tersebut.

![case1](https://github.com/teamalgoritma/community/assets/114390619/6770bd5f-bf13-441c-8e2b-317aa832e823)


Terimakasih.

**Comment by ramdhinafinita** [comment, marked as answer]
Dear @ELITSZORD,

Berikut adalah solusi untuk menjawab kedua problem tersebut:
1. Menyimpan spesifik nama  worksheet:

`data.to_excel("nama_excel.xlsx", sheet_name="nama_sheet", index=False)`

Jika tidak ingin menyertakan indeks sebagai kolom tambahan **Unnamed:0**, Anda dapat menambahkan parameter index=False saat menggunakan to_excel.

2. Setelah melakukan save excel dan dia sudah di set, pastikan kembali apakah semua isi kolom pada date sudah tanggal
 
![image](https://github.com/teamalgoritma/community/assets/64061969/58e23ebb-e1fa-418d-af60-897ea1eed05e)
Contoh pembacaan langsung:
![image](https://github.com/teamalgoritma/community/assets/64061969/f26081be-ca0d-4959-9f0c-326f629398d6)

Jika terdapat kendala pada kolom date masih terubah kedalam bentuk string, alternatif bisa menggunakan paramater parse_dates = True saat read_excel

**Reply by ELITSZORD** [reply]
Terimakasih atas responnya kak @ramdhinafinita dan @arkanayudhistira , apakah jawaban nomor 1 bisa di pakai ketika ada suatu file excel dimana ada bebrapa worksheet (A,B,C,D), nah misal df baru ini mau kita tiban/overwrite kedalam worksheet D tanpa merubah tipe datanya itu gimana ya?

Dan untuk jawaban nomor 2 hasilnya kenapa tidak bisa disaya ya kak..
